timemachine/fe/single_topology_v3.py

```diff
@@ -0,0 +1,539 @@
+            # So instead, we will pick a random anchor atom. One possible solution later on is to
+            # minimize the number of rotatable bonds?
+            warnings.warn(
+                f"Multiple root anchors {root_anchors} found for dummy group: {dg}, picking a random anchor.",
+                MultipleAnchorWarning,
+            )
+
+        jk = None
+
+        # find an arbitrary but stable angle core atom that is one bond away from the root atom that is
+        # present in mol_a
+        for ra in root_anchors:
+            core_b_nbs = [a.GetIdx() for a in mol_b.GetAtomWithIdx(ra).GetNeighbors() if a.GetIdx() in core_b]
+            for nb in core_b_nbs:
+                # see if this core_bond is present in mol_a
+                # first, look up the idxs of the atoms in mol_a:
+                ra_a, nb_a = core_b_to_a[ra], core_b_to_a[nb]
+                if (ra_a, nb_a) in bond_idxs_a or (nb_a, ra_a) in bond_idxs_a:
+                    jk = ra, nb
+                    break
+
+        if jk is None:
+            warnings.warn("Unable to find stable angle term in mol_a", CoreBondChangeWarning)
+            # pick an arbitrary root_anchor
+            jk = root_anchors[0], None
+
+        jks.append(jk)
+
+    return dummy_groups_b, jks
+
+
+class SingleTopologyV3:
+    def __init__(self, mol_a, mol_b, core, forcefield):
+        """
+        SingleTopology combines two molecules through a common core. The combined mol has
+        atom indices laid out such that mol_a is identically mapped to the combined mol indices.
+        The atoms in the mol_b's R-group is then glued on to resulting molecule.
+
+        Parameters
+        ----------
+        mol_a: ROMol
+            First ligand
+        mol_b: ROMol
+            Second ligand
+        core: np.array (C, 2)
+            Atom mapping from mol_a to to mol_b.
+        ff: ff.Forcefield
+            Forcefield to be used for parameterization.
+
+        """
+        assert mol_a is not None
+        assert mol_b is not None
+        self.mol_a = mol_a
+        self.mol_b = mol_b
+        self.core = core
+        self.ff = forcefield
+
+        assert core.shape[1] == 2
+
+        # map into idxs in the combined molecule
+        self.a_to_c = np.arange(mol_a.GetNumAtoms(), dtype=np.int32)  # identity
+        self.b_to_c = np.zeros(mol_b.GetNumAtoms(), dtype=np.int32) - 1
+
+        self.NC = mol_a.GetNumAtoms() + mol_b.GetNumAtoms() - len(core)
+
+        # mark membership in the combined molecule:
+        # 0: CORE
+        # 1: DUMMY_A (default)
+        # 2: DUMMY_B
+        self.c_flags = np.ones(self.get_num_atoms(), dtype=np.int32)
+
+        for a, b in core:
+            self.c_flags[a] = 0
+            self.b_to_c[b] = a
+
+        iota = self.mol_a.GetNumAtoms()
+        for b_idx, c_idx in enumerate(self.b_to_c):
+            if c_idx == -1:
+                self.b_to_c[b_idx] = iota
+                self.c_flags[iota] = 2
+                iota += 1
```

Comment:
Question: Should we warn or log if `c_idx != -1`?

Reply:
edit - loop is still needed to setup `b_to_c`, we do not need to warn or log, it is expected that the `b_to_c` will already be mapped (namely, core atoms in `b`), as shown in the above loop. This code sets up the idxs in the combined molecule for the *dummy* atoms of `b`